### PR TITLE
Added 'withinPolygon' to query

### PR DESCRIPTION
**Pull request by dplewis**
This PR addresses issue #3290 

My first PR any help would be great.

I don't know much about Postgres. Is there a guide to setup Postgres with parse server locally?

```
curl -X POST \
  -H "X-Parse-Application-Id: app-id-here" \
  -H "Content-Type: application/json" \
  -d '{
        "location": {
          "__type": "GeoPoint",
          "latitude": 1.5,
          "longitude": 1.5
        }
      }' \
http://localhost:1337/parse/classes/TestObject
```

```
curl -X GET \
-H "X-Parse-Application-Id: app-id-here" \
-H "Content-Type: application/json" \
  -G \
  --data-urlencode 'where={
        "location": {
          "$geoWithin": {
            "$polygon": [
              {
                "__type": "GeoPoint",
                "latitude": 1,
                "longitude": 1
              },
              {
                "__type": "GeoPoint",
                "latitude": 1,
                "longitude": 2
              },
              {
                "__type": "GeoPoint",
                "latitude": 2,
                "longitude": 2
              },
              {
                "__type": "GeoPoint",
                "latitude": 2,
                "longitude": 1
              }
            ]
          }
        }
      }' \
http://localhost:1337/parse/classes/TestObject
```

### CODE REVIEW

**Comment by flovilmart** [comment]
Hi @dplewis Thanks for this PR! That'S great that you jump in on such requested feature!

Would you please add some unit tests to cover this feature?

The best location would be [here](https://github.com/parse-community/parse-server/blob/master/spec/ParseGeoPoint.spec.js) where other geoPoints tests are run.

If you can add a tests that covers:

1. Bad input (malformed inputs, etc...)
2. Successfully fetches all objects that have points within the polygon, while excluding the ones outside.

Because the JS SDK doesn't have the within polygon yet, you should use request and craft the query directly as a REST request. You can find an example query request [here](https://github.com/parse-community/parse-server/blob/master/spec/ParseQuery.spec.js#L2806)
Thanks!

**Comment by dplewis** [comment]
Test have been added

**Comment by flovilmart** [comment]
Can we add a test with invalid points within the polygon array?

**Comment by dplewis** [comment]
Sorry for the late reply. I've added more tests

**Comment by dplewis** [comment]
Quick question when writing test do the test objects get deleted before and after the Unit Tests

**Comment by flovilmart** [comment]
Yes, the whole DB is cleaned up between tests

**Comment by dplewis** [comment]
Interesting at one point the test passed the first time but second and third, etc failed until I deleted all objects.

**Comment by flovilmart** [comment]
Seems that it's not properly supported on Postgres. Can you check the Travis logs?

**Comment by dplewis** [comment]
How do I go about running tests against Postgres

**Comment by flovilmart** [comment]
You can disable the Postgres test for now with it_exclude_db() (Check in other tests for the exact syntax) 
Or to run the pg tests: 

`PARSE_SERVER_TEST_DB=postgres npm test` 

You'll need Postgres installed and configured (as in the .travis.yaml)

**Comment by dplewis** [comment]
I got the test to pass locally with postgres

**Comment by dplewis** [comment]
Can you create add testing Postgres to `CONTRIBUTING.md`?

I can try to help out on the Postgres Adapter.

**Comment by flovilmart** [comment]
> Can you create add testing Postgres to CONTRIBUTING.md?

I'll work on the contributing guidelines :)

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 1]
```diff
@@ -10,6 +10,7 @@ const PostgresDuplicateObjectError = '42710';
 const PostgresUniqueIndexViolationError = '23505';
 const PostgresTransactionAbortedError = '25P02';
 const logger = require('../../../logger');
+const _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };
```
Comment (flovilmart) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: does that come from transpired babel source? please remove :)

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js [hunk 2]
```diff
@@ -344,6 +345,25 @@ const buildWhereClause = ({ schema, query, index }) => {
       index += 2;
     }
 
+    if (fieldValue.$geoWithin && fieldValue.$geoWithin.$polygon) {
+      const polygon = fieldValue.$geoWithin.$polygon;
+      if (!(polygon instanceof Array)) {
+        throw new Parse.Error(Parse.Error.INVALID_JSON, 'bad $geoWithin value');
+      }
+      const points = polygon.map((point) => {
+        if (!((typeof point === 'undefined' ? 'undefined' : _typeof(point)) === 'object' && point !== null && point.__type === 'GeoPoint')) {
```
Comment (flovilmart) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: doesn't seem to be something a human would write :)
can we replace with:

```
if (typeof point !== 'object' || point.__type !== 'GeoPoint') {
```

**Comment by flovilmart** [comment]
@dplewis just one small nit, otherwise it's looking awesome!

File: src/Adapters/Storage/Postgres/PostgresStorageAdapter.js (hunk 2 above)
Comment (dplewis) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Good thing I'm not human

Comment (flovilmart) [inline src/Adapters/Storage/Postgres/PostgresStorageAdapter.js]: Sorry if I offended you, it looked like generated code

**Comment by dplewis** [comment]
Thanks, when can we update the SDKs?

Also theres a debug function in PostgresAdapter. How do I run that?

**Comment by flovilmart** [comment]
With `PARSE_SERVER_LOG_LEVEL=debug`

**Comment by flovilmart** [comment]
@dplewis we can update the SDK's anytime, on old parse-server versions that will just fail as unsupported queries, with next release, this should work nicely

**Comment by madsb** [comment]
Holy smokes, this looks awesome - can't wait to try it out. Thanks, @dplewis! 🙏🏼

**Comment by otymartin** [comment]
@dplewis will you be updating the iOS SDK with this aswell?

**Comment by dplewis** [comment]
@otymartin I'm working on it now should be there soon

**Comment by montymxb** [comment]
@dplewis nice to see this bumping along! Has the android sdk received this yet by chance?

**Comment by otymartin** [comment]
@dplewis nice thanks man

**Pull request merged**